lua/neo-tree/sources/filesystem/init.lua

```diff
@@ -428,4 +429,76 @@ M.toggle_directory = function(state, node, path_to_reveal, skip_redraw, recursiv
   end
 end
 
+--- Recursively expand all loaded nodes under the given node
+--- returns table with all discovered nodes that need to be loaded 
+---@param node table a node to expand
+---@param state table current state of the source
+---@return table discovered nodes that need to be loaded
+local function expand_loaded(node, state)
+    local function rec(_node, to_load)
+      if _node.loaded == false then
+        table.insert(to_load, _node)
+      else
+        if not _node:is_expanded() then
+          _node:expand()
+          state.explicitly_opened_directories[_node:get_id()] = true
+        end
+        local children = state.tree:get_nodes(_node:get_id())
+        log.debug("Expanding childrens of " .. _node:get_id())
+        for _, child in ipairs(children) do
+          if child.type == "directory" then
+             rec(child, to_load)
+          else
+            log.trace("Child: " .. child.name .. " is not a directory, skipping")
+          end
+        end
+      end
+    end
+
+    local to_load = {}
+    rec(node, to_load)
+    return to_load
+end
+
+--- Recursively expands all nodes under the given node
+--- loading nodes if necessary.
+--- asyn method
+---@param node table a node to expand
+---@param state table current state of the source
+local function expand_and_load(node, state)
+    local function rec(to_load, progress)
+      local to_load_current = expand_loaded(node, state)
+      for _,v in ipairs(to_load_current) do
+        table.insert(to_load, v)
+      end
+      if progress <= #to_load then
+        M.expand_directory(state, to_load[progress])
+        rec(to_load, progress + 1)
+      end
+    end
+    rec({}, 1)
+end
+
+---Expands given node recursively loading all descendant nodes if needed
+---async method
+---@param state table current state of the source
+---@param node table a node to expand
+M.expand_directory = function(state, node)
+  log.debug("Expanding directory " .. node:get_id())
+  if node.type ~= "directory" then
+    return
+  end
+  state.explicitly_opened_directories = state.explicitly_opened_directories or {}
+  if node.loaded == false then
+    local id = node:get_id()
+    state.explicitly_opened_directories[id] = true
+    renderer.position.set(state, nil)
+    fs_scan.get_dir_items_async(state, id, true)
+    -- ignore results as we know here that all descendant nodes have been already loaded
+    expand_loaded(node ,state)
```

Comment:
Here is the problem - for some reason `fs_scan.get_dir_items_async` does set `loaded=true` for all the nodes in the tree despite recursively scanning whole directory. 

`fs_scan.get_dir_items_async` is a simplification of `fs_scan.get_items`

If you could point me to a place that sets `loaded` I will be able to debug it further. 

Right now, expanding `all_nodes` works but is inefficient due to redundant directory scans.

Reply:
Ok, I misunderstood how the `scan_dir_async` and `scan_dir_sync` behave. I thought that since we can pass there `recursive` flag they should be able to load all nodes recursively from the file system, and then all the nodes can be processed (like e.g. by expanding them). 

However, these functions load only one level below and changing them is not so easy as `renderContext` that I wanted to call for each scanned directory to create corresponding nodes in the tree clears all the variables in the context (there might be other obstacles to setting that). 

So, right now it works both for `expand_all` and `expand_under_cursor`. I am just not sure if it was supposed to work like this.

Reply:
@cseickel Could you also confirm that this is how it is supposed to work? 

I have a strong feeling that this is not exactly right, but maybe I still don't understand some things. I looked once again and it seems that this was the original intention as the `async_scan` with `scan_mode==shallow` scans directories recursively deep. So there is no reason to call it then multiple times for every new child.

Reply:
nvm, I solved it, see this commit: https://github.com/nvim-neo-tree/neo-tree.nvim/pull/957/commits/b5a72242925026dba8de79b279ecf71e2c4fd434